tests/org.eclipse.swt.tests/JUnit Tests/org/eclipse/swt/tests/junit/performance/Test_org_eclipse_swt_widgets_Tree.java

```diff
@@ -0,0 +1,359 @@
+			return measureNanos(() -> {
+				int[] count = new int[] { 0 };
+				breadthFirstTraverse(tree, item -> count[0]++);
+				Assert.assertEquals(n, count[0]);
+			});
+		});
+	}
+
+	@Test
+	public void secondTraverse() {
+		assertMaximumDegree(1.2, n -> {
+			Tree tree = buildSubject(n, this::initializeItem);
+			breadthFirstTraverse(tree, item -> {
+				item.setExpanded(true);
+			});
+			return measureNanos(() -> {
+				breadthFirstTraverse(tree, item -> {
+				});
+			});
+		});
+	}
+
+	@Test
+	public void getForeground() {
+		assertMaximumDegree(1.2, n -> {
+			Tree tree = buildSubject(n, this::initializeItem);
+			breadthFirstTraverse(tree, item -> {
+				item.setExpanded(true);
+			});
+			return measureNanos(() -> {
+				breadthFirstTraverse(tree, TreeItem::getForeground);
+			});
+		});
+	}
+
+	@Test
+	public void setForeground() {
+		assertMaximumDegree(1.3, n -> {
+			Tree tree = buildSubject(n, this::initializeItem);
+			breadthFirstTraverse(tree, item -> {
+				item.setExpanded(true);
+			});
+			Color cyan = shell.getDisplay().getSystemColor(SWT.COLOR_CYAN);
+			return measureNanos(() -> {
+				breadthFirstTraverse(tree, item -> item.setForeground(cyan));
+			});
+		});
+	}
+
+	@Test
+	public void setText() {
+		assertMaximumDegree(1.3, n -> {
+			Tree tree = buildSubject(n, this::initializeItem);
+			breadthFirstTraverse(tree, item -> {
+				item.setExpanded(true);
+			});
+			return measureNanos(() -> {
+				breadthFirstTraverse(tree, item -> item.setText("test"));
+			});
+		});
+	}
+
+	@Test
+	public void showItem() {
+		assertMaximumDegree(virtual ? 1.2 : 1.9, n -> {
+			Tree tree = buildSubject(n, this::initializeItem);
+			return measureNanos(() -> tree.showItem(shape.lastItem(tree)));
+		});
+	}
+
+	@Test
+	public void jfaceReveal() {
+		// Still demonstrates different characteristics
+		assertMaximumDegree(virtual ? 1.1 : 2, n -> {
+			Tree tree = buildSubject(n, this::initializeItem);
+			return measureNanos(() -> {
+				// JFace creates all children and updates each by its index
+
+				TreeItem item = tree.getItem(tree.getItemCount() - 1);
+
+				for (;;) {
+					item.setExpanded(true);
+					int count = item.getItemCount();
+					if (virtual) {
+						item.getItems();
+						for (int i = 0; i < count; i++) {
+							item.getItem(i); // JFace updates children using their indices
+						}
+					}
+					if (count <= 0) {
+						tree.showItem(item);
+						break;
+					}
+					item = item.getItem(count - 1);
+				}
+			});
+		});
+	}
+
+	private Tree buildSubject(int size, Consumer<TreeItem> initialize) {
+		Tree result = new Tree(shell, virtual ? SWT.VIRTUAL : SWT.NONE);
+		shell.layout();
+		result.setRedraw(false);
+		shape.buildTree(result, size, initialize);
+		result.setRedraw(true);
+		return result;
+	}
+
+	/** Ensure that given function grows within acceptable polynomial degree */
+	private void assertMaximumDegree(double maximumDegree, IntFunction<Double> function) {
```

Comment:
Asserting on times is unreliable.
If system experience sudden slowdown for a second, tests will fail.
It's fine however if tests are designed to be run manually only.

The other problem is that you're asserting complexity for ALL platforms, not just GTK.
To me it makes sense to convert JUnit test to a manual snippet.

Reply:
I'm trying to fix a performance problem. Note that I do not assert the absolute time and instead check the ratios. There is no way to prevent performance regression besides running the tests in CI. I specifically enabled these tests on CI to be able to have a single point of reference and universal source of timings for the process of this review. The tests are designed to be run on CI.

Checking all the platforms is intentional, as the problem I'm trying to fix affects only Linux and tests should demonstrate it. No platform should exhibit quadratic behaviors, so they also serve as regression tests.

There are already manual snippets in the repository testing performance, but they failed to find the problem in question. I assume nobody runs them.

Reply:
Test is now removed from suites and stays passive next to other abandoned performance tests.